### ISSUE TITLE
Unexpected high memory consumption when typing recursive functions

### ISSUE THREAD

**Issue by jkjkil4** [Myriad-Dreamin/tinymist#754]
**Describe the bug**
Memory leak when typing in editor.

TL;DR
- typing `#fib(12)` works well
- typing `#fib(56)` could result in memory leak
- type `#fib()`, click outside the brackets, click back and complete `#fib(56)`, works well 

To reproduce:

```typ
#let fib(n) = {
    if n <= 2 {
        1
    } else {
        fib(n - 1) + fib(n - 2)
    }
}

#fib(56)
```

when typing the last line directly like:

```typ
#fib
```

```typ
#fib()
```

```typ
#fib(5)
```

```typ
#fib(56)
```

then the tinymist process could consume large amount of RAM and get stuck:

![image](https://github.com/user-attachments/assets/039f735b-9483-4b81-96f1-db4aae01cced)

(got the result of `#fib(56)` but tinymist refuse to update any other changes after a while)

I didn't find this problem when typing `#fib(12)`

But, when typing the last line separately like:

```typ
#fib
```

```typ
#fib()
```

click outside the brackets, then click back and continue

```typ
#fib(5)
```

```typ
#fib(56)
```

it works well and won't cause memory leak.

**Package/Software version:**

VSCode version(Help -> About):
```plain
Version: 1.95.0 (user setup)
Commit: 912bb683695358a54ae0c670461738984cbb5b95
Date: 2024-10-28T20:16:24.561Z
Electron: 32.2.1
ElectronBuildId: 10427718
Chromium: 128.0.6613.186
Node.js: 20.18.0
V8: 12.8.374.38-electron.0
OS: Windows_NT x64 10.0.19045
```

tinymist extension version: `v0.12.0`. Get it by `tinymist --version` in terminal.
```plain
I couldn't get this section, skipped.
```

**Logs:**

tinymist server log(Output Panel -> tinymist):
```plain
[2024-10-31T00:51:49Z INFO  sync_lsp] handling workspace/executeCommand - (36) at Instant { t: 3130.40s }
[2024-10-31T00:51:49Z INFO  sync_lsp] handled  workspace/executeCommand - (36) in 74.80µs
[2024-10-31T00:51:49Z INFO  sync_lsp] handling textDocument/codeLens - (37) at Instant { t: 3130.71s }
[2024-10-31T00:51:49Z INFO  sync_lsp] handled  textDocument/codeLens - (37) in 507.00µs
[2024-10-31T00:51:50Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3131.37s }
[2024-10-31T00:51:50Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 251.10µs
[2024-10-31T00:51:50Z INFO  tinymist::actor::editor] received status request(primary) Compiling
error: expected expression
   ┌─ d:\Projects\typst\test.typ:10:1
   │
10 │ #
   │  ^

[2024-10-31T00:51:50Z INFO  tinymist::actor::editor] received diagnostics from primary:19: diag(Some(1))
[2024-10-31T00:51:50Z INFO  tinymist::actor::editor] received status request(primary) CompileError
[2024-10-31T00:51:50Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 34.5µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/completion - (38) at Instant { t: 3131.38s }
[2024-10-31T00:51:50Z INFO  tinymist_query::upstream::complete] continue after completing type
[2024-10-31T00:51:50Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 250.5µs
[2024-10-31T00:51:50Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 336.5µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/completion - (38) in 35.53ms
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/inlayHint - (39) at Instant { t: 3131.47s }
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/inlayHint - (39) in 228.70µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/foldingRange - (40) at Instant { t: 3131.57s }
[2024-10-31T00:51:50Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 155.8µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/foldingRange - (40) in 362.20µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/codeLens - (41) at Instant { t: 3131.63s }
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/codeLens - (41) in 528.20µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/semanticTokens/full/delta - (42) at Instant { t: 3131.67s }
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/semanticTokens/full/delta - (42) in 289.30µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/codeAction - (43) at Instant { t: 3131.71s }
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/codeAction - (43) in 323.60µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/documentSymbol - (44) at Instant { t: 3131.75s }
[2024-10-31T00:51:50Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 181.9µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/documentSymbol - (44) in 733.80µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handling textDocument/documentSymbol - (45) at Instant { t: 3131.75s }
[2024-10-31T00:51:50Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 167.5µs
[2024-10-31T00:51:50Z INFO  sync_lsp] handled  textDocument/documentSymbol - (45) in 411.00µs
[2024-10-31T00:51:51Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3131.85s }
[2024-10-31T00:51:51Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 310.90µs
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) Compiling
error: unknown variable: f
   ┌─ d:\Projects\typst\test.typ:10:1
   │
10 │ #f
   │  ^

[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received diagnostics from primary:20: diag(Some(1))
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) CompileError
[2024-10-31T00:51:51Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 41.2µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/inlayHint - (46) at Instant { t: 3131.92s }
[2024-10-31T00:51:51Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 89.9µs
[2024-10-31T00:51:51Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 219.5µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/inlayHint - (46) in 956.00µs
[2024-10-31T00:51:51Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3131.97s }
[2024-10-31T00:51:51Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 263.80µs
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) Compiling
error: unknown variable: fi
   ┌─ d:\Projects\typst\test.typ:10:1
   │
10 │ #fi
   │  ^^

[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received diagnostics from primary:21: diag(Some(1))
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) CompileError
[2024-10-31T00:51:51Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 42.3µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/inlayHint - (47) at Instant { t: 3132.08s }
[2024-10-31T00:51:51Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 153.6µs
[2024-10-31T00:51:51Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 203.9µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/inlayHint - (47) in 965.30µs
[2024-10-31T00:51:51Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3132.17s }
[2024-10-31T00:51:51Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 269.70µs
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) Compiling
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received diagnostics from primary:22: diag(Some(0))
[2024-10-31T00:51:51Z DEBUG typst_preview::actor::render] OutlineRenderActor: received message: RenderIncremental
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) CompileSuccess
[2024-10-31T00:51:51Z DEBUG typst_preview::actor::render] OutlineRenderActor: sending outline
[2024-10-31T00:51:51Z DEBUG typst_preview::actor::render] OutlineRenderActor: waiting for message
[2024-10-31T00:51:51Z DEBUG typst_preview::actor::render] RenderActor: has_full_render: false
[2024-10-31T00:51:51Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 54.5µs
[2024-10-31T00:51:51Z DEBUG typst_preview::actor::render] RenderActor: waiting for message
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/semanticTokens/full/delta - (48) at Instant { t: 3132.22s }
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/semanticTokens/full/delta - (48) in 380.60µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/inlayHint - (49) at Instant { t: 3132.23s }
[2024-10-31T00:51:51Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 69.1µs
[2024-10-31T00:51:51Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 173.1µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/inlayHint - (49) in 947.90µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/foldingRange - (50) at Instant { t: 3132.39s }
[2024-10-31T00:51:51Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 101.4µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/foldingRange - (50) in 357.50µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/documentColor - (51) at Instant { t: 3132.39s }
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/documentColor - (51) in 344.40µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/codeLens - (52) at Instant { t: 3132.44s }
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/codeLens - (52) in 531.80µs
[2024-10-31T00:51:51Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3132.45s }
[2024-10-31T00:51:51Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 315.70µs
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) Compiling
error: missing argument: n
   ┌─ d:\Projects\typst\test.typ:10:1
   │
10 │ #fib()
   │  ^^^^^

[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received diagnostics from primary:23: diag(Some(1))
[2024-10-31T00:51:51Z INFO  tinymist::actor::editor] received status request(primary) CompileError
[2024-10-31T00:51:51Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 61.5µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/completion - (53) at Instant { t: 3132.47s }
[2024-10-31T00:51:51Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 140µs
[2024-10-31T00:51:51Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 232.6µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/completion - (53) in 30.22ms
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/inlayHint - (54) at Instant { t: 3132.53s }
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/inlayHint - (54) in 511.70µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/signatureHelp - (55) at Instant { t: 3132.59s }
[2024-10-31T00:51:51Z INFO  tinymist_query::analysis::linked_def] def_name for function: Ident: "fib"
[2024-10-31T00:51:51Z INFO  tinymist_query::analysis::linked_def] okay for function: Some((Func(fib), None))
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/signatureHelp - (55) in 342.00µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/foldingRange - (56) at Instant { t: 3132.67s }
[2024-10-31T00:51:51Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 54.4µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/foldingRange - (56) in 193.80µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/codeLens - (57) at Instant { t: 3132.72s }
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/codeLens - (57) in 547.20µs
[2024-10-31T00:51:51Z INFO  sync_lsp] handling textDocument/semanticTokens/full/delta - (58) at Instant { t: 3132.76s }
[2024-10-31T00:51:51Z INFO  sync_lsp] handled  textDocument/semanticTokens/full/delta - (58) in 363.90µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/documentSymbol - (59) at Instant { t: 3132.85s }
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 73.5µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/documentSymbol - (59) in 322.10µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/codeAction - (60) at Instant { t: 3132.86s }
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/documentSymbol - (61) at Instant { t: 3132.86s }
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/codeAction - (60) in 791.20µs
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 47.1µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/documentSymbol - (61) in 197.70µs
[2024-10-31T00:51:52Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3133.02s }
[2024-10-31T00:51:52Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 395.00µs
[2024-10-31T00:51:52Z INFO  tinymist::actor::editor] received status request(primary) Compiling
[2024-10-31T00:51:52Z INFO  tinymist::actor::editor] received diagnostics from primary:24: diag(Some(0))
[2024-10-31T00:51:52Z INFO  tinymist::actor::editor] received status request(primary) CompileSuccess
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] OutlineRenderActor: received message: RenderIncremental
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] RenderActor: has_full_render: false
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] OutlineRenderActor: sending outline
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] OutlineRenderActor: waiting for message
[2024-10-31T00:51:52Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 43.5µs
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] RenderActor: waiting for message
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/inlayHint - (62) at Instant { t: 3133.08s }
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 133.2µs
[2024-10-31T00:51:52Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 223.2µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/inlayHint - (62) in 944.40µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/signatureHelp - (63) at Instant { t: 3133.18s }
[2024-10-31T00:51:52Z INFO  tinymist_query::analysis::linked_def] def_name for function: Ident: "fib"
[2024-10-31T00:51:52Z INFO  tinymist_query::analysis::linked_def] okay for function: Some((Func(fib), None))
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/signatureHelp - (63) in 1.22ms
[2024-10-31T00:51:52Z INFO  sync_lsp] notifying textDocument/didChange - at Instant { t: 3133.24s }
[2024-10-31T00:51:52Z INFO  sync_lsp] notifing textDocument/didChange succeeded in 251.70µs
[2024-10-31T00:51:52Z INFO  tinymist::actor::editor] received status request(primary) Compiling
[2024-10-31T00:51:52Z INFO  tinymist::actor::editor] received diagnostics from primary:25: diag(Some(0))
[2024-10-31T00:51:52Z INFO  tinymist::actor::editor] received status request(primary) CompileSuccess
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] OutlineRenderActor: received message: RenderIncremental
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] RenderActor: has_full_render: false
[2024-10-31T00:51:52Z INFO  tinymist::task::cache] CacheEvictTask: evict cache in 36.1µs
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] OutlineRenderActor: sending outline
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] OutlineRenderActor: waiting for message
[2024-10-31T00:51:52Z DEBUG typst_preview::actor::render] RenderActor: waiting for message
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/inlayHint - (64) at Instant { t: 3133.29s }
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 70.4µs
[2024-10-31T00:51:52Z INFO  tinymist_query::analysis::tyck] Type checking on \test.typ took 172.4µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/inlayHint - (64) in 938.70µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/semanticTokens/full/delta - (65) at Instant { t: 3133.36s }
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/semanticTokens/full/delta - (65) in 347.40µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/signatureHelp - (66) at Instant { t: 3133.39s }
[2024-10-31T00:51:52Z INFO  tinymist_query::analysis::linked_def] def_name for function: Ident: "fib"
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/foldingRange - (67) at Instant { t: 3133.46s }
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 94.5µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/foldingRange - (67) in 311.80µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/documentColor - (68) at Instant { t: 3133.47s }
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/documentColor - (68) in 561.30µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/codeAction - (69) at Instant { t: 3133.51s }
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/codeAction - (69) in 314.80µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/codeLens - (70) at Instant { t: 3133.52s }
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/codeLens - (70) in 689.90µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/documentSymbol - (71) at Instant { t: 3133.62s }
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 93.6µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/documentSymbol - (71) in 337.10µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handling textDocument/documentSymbol - (72) at Instant { t: 3133.62s }
[2024-10-31T00:51:52Z INFO  tinymist_query::syntax::lexical_hierarchy] lexical hierarchy analysis took 116.6µs
[2024-10-31T00:51:52Z INFO  sync_lsp] handled  textDocument/documentSymbol - (72) in 322.80µs
[2024-10-31T00:51:53Z INFO  sync_lsp] handling textDocument/inlayHint - (73) at Instant { t: 3134.06s }
[2024-10-31T00:51:53Z INFO  sync_lsp] handled  textDocument/inlayHint - (73) in 180.40µs
[2024-10-31T00:51:53Z INFO  sync_lsp] handling textDocument/documentLink - (74) at Instant { t: 3134.26s }
[2024-10-31T00:51:53Z INFO  sync_lsp] handled  textDocument/documentLink - (74) in 187.10µs
[2024-10-31T00:51:55Z INFO  sync_lsp] handling workspace/executeCommand - (75) at Instant { t: 3136.02s }
[2024-10-31T00:51:55Z INFO  sync_lsp] handled  workspace/executeCommand - (75) in 33.30µs
[2024-10-31T00:51:55Z INFO  sync_lsp] notifying $/cancelRequest - at Instant { t: 3136.03s }
[2024-10-31T00:51:55Z WARN  sync_lsp] unhandled notification: $/cancelRequest
```

tinymist client log(Help -> Toggle Developer Tools -> Console):
```plain
VM17:5926 batch 1 messages
VM17:5947 recv diff-v1 1404
VM17:1388 render_in_window with partial rendering enabled window 0.8186813186813187 0 0.8186813186813187 609.0989010989011 245.6043956043956 , patch scale 2023
VM17:369 parse 2.50 ms, rerender 19.70 ms, total 22.20 ms
VM17:1832 updateCanvas start
VM17:1872 updateCanvas done 0.39999999990686774
VM17:5926 batch 1 messages
VM17:5947 recv diff-v1 908
VM17:1388 render_in_window with partial rendering enabled window 0.8186813186813187 0 0.8186813186813187 609.0989010989011 245.6043956043956 , patch scale 1489
VM17:369 parse 0.30 ms, rerender 5.10 ms, total 5.40 ms
VM17:1832 updateCanvas start
VM17:1872 updateCanvas done 0.5
VM17:5926 batch 1 messages
VM17:5947 recv diff-v1 3220
VM17:1388 render_in_window with partial rendering enabled window 0.8186813186813187 0 0.8186813186813187 609.0989010989011 245.6043956043956 , patch scale 4145
VM17:369 parse 0.40 ms, rerender 4.50 ms, total 4.90 ms
VM17:1832 updateCanvas start
VM17:1872 updateCanvas done 0.2999999998137355
console.ts:137 [Extension Host] skip extension-output file extension-output-myriad-dreamin.tinymist-#1-Tinymist Typst Language Server
```

**Additional context**
- It also happens without the preview page
- I also reproduced this issue on my roommate's device.

**Comment by Enter-tainer** [comment]
Will `typst watch` consume a lot of ram in this case?

**Comment by jkjkil4** [comment]
It doesn't happen when using `typst watch`.
I guess it is unrelated, as it could still occur even without the preview page, just by using the tinymist syntax hints.

**Comment by Enter-tainer** [comment]
> It doesn't happen when using typst watch.
> I guess it is unrelated, as it could still occur even without the preview page, just by using the tinymist syntax hints.

Thanks, this help us triage this issue.

**Comment by jkjkil4** [comment]
I noticed that for smaller `n`, tinymist can also consume a lot of RAM.
In these cases, I can measure the RAM usage since it is much lower than `#fib(56)`:
- `#fib(29)`: 446.7MB
- `#fib(30)`: 716.3MB
- `#fib(31)`: 1152.8MB
- `#fib(32)`: 1855.5MB
- `#fib(33)`: 2994.5MB
(I restart tinymist before these measurements to ensure they are independent.)

I know this Fibonacci sequence algorithm is exponential, but it's strange that doing syntax hints for it could result in high memory consumption; this may explain why tinymist gets stuck in the case of `#fib(56)`."
